### PR TITLE
Avoid lambda compilation for member access expressions in LINQ

### LINKED ISSUE #2947: LINQ queries triggers JIT a bit too much
I was profiling our CPU-heavy application and apparently there are a lot of traces where jit is invoked to compile some lambdas when issuing a LINQ `ISession.Query<TRecord>`

This is noticable especially when using `dotnet counters monitor -p1`, example:
```
[System.Runtime]
    % Time in GC since last GC (%)                                 8    
    Allocation Rate (B / 1 sec)                               3.2411e+08   
    CPU Usage (%)                                                 69    
    Exception Count (Count / 1 sec)                                9    
    GC Committed Bytes (MB)                                    3,389    
    GC Fragmentation (%)                                          27.191
    GC Heap Size (MB)                                          2,344    
    Gen 0 GC Count (Count / 1 sec)                                 0    
    Gen 0 Size (B)                                               192    
    Gen 1 GC Count (Count / 1 sec)                                 0    
    Gen 1 Size (B)                                            1.1532e+08   
    Gen 2 GC Count (Count / 1 sec)                                 0    
    Gen 2 Size (B)                                            8.8658e+08   
    IL Bytes Jitted (B)                                       2.3988e+08   
    LOH Size (B)                                              4.3898e+08   
    Monitor Lock Contention Count (Count / 1 sec)                116    
    Number of Active Timers                                      312    
    Number of Assemblies Loaded                                  305    
    Number of Methods Jitted                               8,407,800    
    POH (Pinned Object Heap) Size (B)                        898,696    
    ThreadPool Completed Work Item Count (Count / 1 sec)       7,637    
    ThreadPool Queue Length                                        0    
    ThreadPool Thread Count                                       20    
    Time spent in JIT (ms / 1 sec)                               639.439
    Working Set (MB)                                           3,943    
```

Note `Number of Methods Jitted` and `Time spent in JIT (ms / 1 sec)`

The traces point to `NhPartialEvaluatingExpressionVisitor.EvaluateSubtree` as the culprit. See:
- https://github.com/nhibernate/nhibernate-core/blob/47992e73ed6114093f31ed99829dd8c4c7663829/src/NHibernate/Linq/Visitors/NhPartialEvaluatingExpressionVisitor.cs#L147-L153

Stack trace example where application code looks like:

```
                var record = await db.Query<TRecord>().SingleOrDefaultAsync(
                    s => s.ColumnA == entityA
                    && s.ColumnB == entityB);
```

```
Thread (0x71E):
  [Native Frames]
  System.Private.CoreLib!System.Reflection.Emit.DynamicMethod.CreateDelegate(class System.Type,class System.Object)
  System.Linq.Expressions!System.Linq.Expressions.Compiler.LambdaCompiler.Compile(class System.Linq.Expressions.LambdaExpression)
  NHibernate!NHibernate.Linq.Visitors.NhPartialEvaluatingExpressionVisitor.EvaluateSubtree(class System.Linq.Expressions.Expression)
  NHibernate!NHibernate.Linq.Visitors.NhPartialEvaluatingExpressionVisitor.Visit(class System.Linq.Expressions.Expression)
  System.Linq.Expressions!System.Linq.Expressions.ExpressionVisitor.VisitBinary(class System.Linq.Expressions.BinaryExpression)
  NHibernate!NHibernate.Linq.Visitors.NhPartialEvaluatingExpressionVisitor.Visit(class System.Linq.Expressions.Expression)
  System.Linq.Expressions!System.Linq.Expressions.ExpressionVisitor.VisitBinary(class System.Linq.Expressions.BinaryExpression)
  NHibernate!NHibernate.Linq.Visitors.NhPartialEvaluatingExpressionVisitor.Visit(class System.Linq.Expressions.Expression)
  System.Linq.Expressions!System.Linq.Expressions.ExpressionVisitor.VisitLambda(class System.Linq.Expressions.Expression`1<!!0>)
  NHibernate!NHibernate.Linq.Visitors.NhPartialEvaluatingExpressionVisitor.Visit(class System.Linq.Expressions.Expression)
  System.Linq.Expressions!System.Linq.Expressions.ExpressionVisitor.VisitUnary(class System.Linq.Expressions.UnaryExpression)
  NHibernate!NHibernate.Linq.Visitors.NhPartialEvaluatingExpressionVisitor.Visit(class System.Linq.Expressions.Expression)
  System.Linq.Expressions!System.Dynamic.Utils.ExpressionVisitorUtils.VisitArguments(class System.Linq.Expressions.ExpressionVisitor,class System.Linq.Expressions.IArgumentProvider)
  System.Linq.Expressions!System.Linq.Expressions.ExpressionVisitor.VisitMethodCall(class System.Linq.Expressions.MethodCallExpression)
  NHibernate!NHibernate.Linq.Visitors.NhPartialEvaluatingExpressionVisitor.Visit(class System.Linq.Expressions.Expression)
  NHibernate!NHibernate.Linq.NhRelinqQueryParser.PreTransform(class System.Linq.Expressions.Expression,class NHibernate.Linq.Visitors.PreTransformationParameters)
  NHibernate!NHibernate.Linq.NhLinqExpression..ctor(value class NHibernate.Linq.QueryMode,class System.Linq.Expressions.Expression,class NHibernate.Engine.ISessionFactoryImplementor)
  NHibernate!NHibernate.Linq.DefaultQueryProvider.PrepareQuery(class System.Linq.Expressions.Expression,class NHibernate.IQuery&)
  NHibernate!NHibernate.Linq.DefaultQueryProvider.ExecuteAsync(class System.Linq.Expressions.Expression,value class System.Threading.CancellationToken)
  NHibernate!NHibernate.Linq.DefaultQueryProvider+<ExecuteAsync>d__25`1[System.__Canon].MoveNext()
  System.Private.CoreLib!System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start(!!0&)
  NHibernate!NHibernate.Linq.DefaultQueryProvider.ExecuteAsync(class System.Linq.Expressions.Expression,value class System.Threading.CancellationToken)
  NHibernate!NHibernate.Linq.LinqExtensionMethods.SingleOrDefaultAsync(class System.Linq.IQueryable`1<!!0>,class System.Linq.Expressions.Expression`1<class System.Func`2<!!0,bool>>,value class System.Threading.CancellationToken)

(( private application code/scrubbed ))

  System.Private.CoreLib!System.Threading.Tasks.TaskSchedulerAwaitTaskContinuation+<>c.<Run>b__2_0(class System.Object)
  System.Private.CoreLib!System.Threading.Tasks.Task.ExecuteWithThreadLocal(class System.Threading.Tasks.Task&,class System.Threading.Thread)
  System.Private.CoreLib!System.Threading.Tasks.TaskScheduler.TryExecuteTask(class System.Threading.Tasks.Task)
  Akka!Akka.Actor.ActorCell.HandleActorTaskSchedulerMessage(class Akka.Dispatch.SysMsg.ActorTaskSchedulerMessage)
  Akka!Akka.Actor.ActorCell.SysMsgInvokeAll(value class Akka.Dispatch.SysMsg.EarliestFirstSystemMessageList,int32)
  Akka!Akka.Dispatch.Mailbox.ProcessAllSystemMessages()
  Akka!Akka.Dispatch.Mailbox.<Run>b__36_0()
  Akka!Akka.Actor.ActorCell.UseThreadContext(class System.Action)
  Akka!Akka.Dispatch.Mailbox.Run()
  System.Private.CoreLib!System.Threading.QueueUserWorkItemCallback.Execute()
  System.Private.CoreLib!System.Threading.ThreadPoolWorkQueue.Dispatch()
  System.Private.CoreLib!System.Threading.PortableThreadPool+WorkerThread.WorkerThreadStart()
```

Is this performance/jit profile intended, maybe people have workarounds?

### PR DESCRIPTION
**Pull request by bahusoid**
Fixes #2947 

Also see #2448 for details and benchmarks.

### CODE REVIEW

File: src/NHibernate/Linq/Visitors/NhPartialEvaluatingExpressionVisitor.cs [hunk 1]
```diff
@@ -143,13 +145,14 @@ private Expression EvaluateSubtree(Expression subtree)
...
-				Expression<Func<object>> lambdaWithoutParameters = Expression.Lambda<Func<object>>(Expression.Convert(subtree, typeof(object)));
-				var compiledLambda = lambdaWithoutParameters.Compile();
 
-				object value = compiledLambda();
-				return Expression.Constant(value, subtree.Type);
+			try
+			{
+				return Expression.Constant(ExpressionProcessor.FindValue(subtree), subtree.Type);
+			}
+			catch (TargetInvocationException ex)
```
Comment (bahusoid) [inline src/NHibernate/Linq/Visitors/NhPartialEvaluatingExpressionVisitor.cs:153]: Original code wraps expression in additional lambda to make it compile directly in `Func<object>` delegate.
But my tests show that direct expression compilation and DynamicInvoke works slightly faster. So just unwrap `TargetInvocationException` thrown by `DynamicInvoke` to behave as old code

```
BenchmarkDotNet=v0.13.1, OS=Windows 10.0.19043.1348 (21H1/May2021Update)
Intel Core i5-5200U CPU 2.20GHz (Broadwell), 1 CPU, 4 logical and 2 physical cores
.NET SDK=5.0.201
  [Host]     : .NET 5.0.4 (5.0.421.11614), X64 RyuJIT
  DefaultJob : .NET 5.0.4 (5.0.421.11614), X64 RyuJIT


|        Method |       Mean |     Error |    StdDev | Ratio | RatioSD |
|-------------- |-----------:|----------:|----------:|------:|--------:|
| DynamicInvoke | 183.670 us | 3.5535 us | 3.8022 us | 1.000 |    0.00 |
| WrapAndInvoke | 203.600 us | 1.3428 us | 1.1213 us | 1.106 |    0.02 |
|    Reflection |   1.020 us | 0.0078 us | 0.0070 us | 0.006 |    0.00 |
```

Source: https://gist.github.com/bahusoid/c68cb6b81f8dde9252d0cc67229136b4

**Comment by razzmatazz** [comment]
Hey @bahusoid 

Thanks! Still, some questions:
- Will this will come with 6.x and not on 5.x series? (i.e. a long time later?)
- Will there be a away to backport it?

**Comment by bahusoid** [comment]
> Will this will come with 6.x

It's intended for 5.4. When merged you will be able to test/use it with our dev builds:  https://www.myget.org/gallery/nhibernate
AFAIK it's in pretty stable state right now

File: src/NHibernate/Linq/Visitors/NhPartialEvaluatingExpressionVisitor.cs (hunk 1 above)
Comment (hazzik) [reply src/NHibernate/Linq/Visitors/NhPartialEvaluatingExpressionVisitor.cs:153]: `Reflection` what is it?

**Review by hazzik — APPROVED** [review]

File: src/NHibernate/Linq/Visitors/NhPartialEvaluatingExpressionVisitor.cs (hunk 1 above)
Comment (hazzik) [reply src/NHibernate/Linq/Visitors/NhPartialEvaluatingExpressionVisitor.cs:153]: @bahusoid on my setup WrapAndInvoke is faster than DynamicInvoke

Comment (bahusoid) [reply src/NHibernate/Linq/Visitors/NhPartialEvaluatingExpressionVisitor.cs:153]: Hm... Interesting.. Maybe it's not about setup but target framework?

**Pull request merged**